### PR TITLE
[serve][docs] Add user guide for application builders 

### LINKED ISSUE #34542: [RFC][serve] Serve application arguments
This RFC outlines a proposal to add support for arguments passed to Serve applications. This can be used for iterative development as well as parametrizing applications in production (e.g., use a different model path).

## Problem

Currently it’s difficult to parametrize Serve applications. Take the simple case where I have a wrapper that takes a model path, loads it, and calls it for inference:

```python
@serve.deployment
class ModelWrapper:
    def __init__(self, model_uri: str):
        self._model = load_model(model_uri)

    def __call__(self, *args):
        return self._model(*args)

app = ModelWrapper.bind("s3://model_uri_1")
```

If I want to pass a different model URI to this application (for iterative development or in production to update to a new version), there are currently three options:

- Modify the code itself to have the new path. Problem: I shouldn’t need to update my code to pass an argument!
- Define a `reconfigure` method, build a Serve config file `user_config`. Problem: This doesn’t fit into the `serve run` development workflow and it’s a lot of steps/code for something simple.
- Use environment variables and implement parsing/validation logic myself. This is an OK workaround but it's not very elegant and limits our ability to improve the UX (e.g., show the arguments in the dashboard).

## Proposal
In addition to supporting importing a built application object as we currently do, support builder functions that return a built app. These can take keyword arguments passed via the CLI or config file.

The motivation for a “builder” function is to easily extend to model composition: the flattened arguments can apply to any of the deployments in the application.

CLI usage:
```bash
serve run my_module:build_app model_uri="s3://model_uri_1"
# Easily iterate w/ different arguments.
serve run my_module:build_app model_uri="s3://model_uri_2"
```

Config usage:
```yaml
applications:
  my_app:
    import_path: my_module.build_app
    args:
      model_uri: s3://model_uri_1
```

Python API:
```python
# By default receive a dictionary containing passed args.
def build_app(args: Dict[str, str]) -> BoundDeployment:
    return ModelWrapper.bind(args["model_uri"])

# This can easily be extended to model composition:
def build_composed_app(args: Dict[str, str]) -> BoundDeployment:
    model = ModelWrapper.bind(args["model_uri"])
    return Preprocessor.bind(args["preprocessor_uri"], model)

# Typing support w/ pydantic:
from pydantic import BaseModel

class MyTypedArgs(BaseModel):
    model_uri: str
    model_num_cpus: int

def build_app_typed(args: MyTypedArgs) -> BoundDeployment:
    return Model.options(
        num_replicas=args.model_num_cpus
    ).bind(args.model_uri)
```

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Docs follow-up for: https://github.com/ray-project/ray/pull/34584

## Related issue number

Closes https://github.com/ray-project/ray/issues/34542

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/app-builder-guide.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# Passing Arguments to Applications
+
+This section describes how to pass arguments to your applications using an application builder function.
```
Comment (richardliaw) [inline doc/source/serve/app-builder-guide.md:3]: One concern I have is that we never introduce/formally define the concept of "applications" to users (in key concepts)

Comment (edoakes) [reply doc/source/serve/app-builder-guide.md:3]: you're totally right, I have it on my TODO list to fix that (the getting started guide is updated to reflect this verbiage already).

let me see if I can get that change in for 2.5 as well, should be straightforward.

Comment (angelinalg) [reply doc/source/serve/app-builder-guide.md:3]: Could we add an entry to the Glossary, too? https://docs.ray.io/en/latest/ray-references/glossary.html

File: doc/source/serve/app-builder-guide.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+## Defining an application builder
+
+When writing an application, there are often parameters that you want to be able to easily change in development or production.
+For example, you might have a path to trained model weights and want to test out a newly trained model.
+In Ray Serve, these parameters are typically passed to the constructor of your deployments using `.bind()`.
+This pattern allows deployments to be configured using ordinary Python code but it requires modifying the code anytime one of the parameters needs to change.
```
Comment (angelinalg) [inline doc/source/serve/app-builder-guide.md:10]:
```suggestion
This pattern allows you to configure deployments using ordinary Python code but it requires modifying the code anytime one of the parameters needs to change.
```

File: doc/source/serve/app-builder-guide.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+To pass arguments without changing the code, define an "application builder" function that takes an arguments dictionary (or [Pydantic object](typed-app-builders)) and returns the built application to be run.
+
+```{literalinclude} ../serve/doc_code/app_builder.py
+:start-after: __begin_untyped_builder__
+:end-before: __end_untyped_builder__
+:language: python
+```
+
+This application builder function can be used as the import path in the `serve run` CLI command or the config file (as shown below).
```
Comment (angelinalg) [inline doc/source/serve/app-builder-guide.md:20]:
```suggestion
You can use this application builder function as the import path in the `serve run` CLI command or the config file (as shown below).
```

File: doc/source/serve/app-builder-guide.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+2023-05-16 10:47:31,641 INFO scripts.py:404 -- Running import path: 'hello:app_builder'.
+2023-05-16 10:47:33,344 INFO worker.py:1615 -- Started a local Ray instance. View the dashboard at http://127.0.0.1:8265
+(ServeController pid=56826) INFO 2023-05-16 10:47:35,115 controller 56826 deployment_state.py:1244 - Deploying new version of deployment default_HelloWorld.
+(ServeController pid=56826) INFO 2023-05-16 10:47:35,141 controller 56826 deployment_state.py:1483 - Adding 1 replica to deployment default_HelloWorld.
+(HTTPProxyActor pid=56828) INFO:     Started server process [56828]
+(ServeReplica:default_HelloWorld pid=56830) Message: Hello from CLI
+2023-05-16 10:47:36,131 SUCC scripts.py:424 -- Deployed Serve app successfully.
+```
+
+Notice that the "Hello from CLI" message was printed from within the deployment constructor.
```
Comment (angelinalg) [inline doc/source/serve/app-builder-guide.md:45]:
```suggestion
Notice that the "Hello from CLI" message is printed from within the deployment constructor.
```

File: doc/source/serve/app-builder-guide.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+2023-05-16 10:49:28,678 SUCC scripts.py:419 -- Submitted deploy config successfully.
+(ServeController pid=57109) INFO 2023-05-16 10:49:28,676 controller 57109 controller.py:559 - Starting deploy_serve_application task for application MyApp.
+(HTTPProxyActor pid=57111) INFO:     Started server process [57111]
+(ServeController pid=57109) INFO 2023-05-16 10:49:28,942 controller 57109 deployment_state.py:1244 - Deploying new version of deployment MyApp_HelloWorld.
+(ServeController pid=57109) INFO 2023-05-16 10:49:29,016 controller 57109 deployment_state.py:1483 - Adding 1 replica to deployment MyApp_HelloWorld.
+(ServeReplica:MyApp_HelloWorld pid=57113) Message: Hello from config
+(ServeController pid=57109) INFO 2023-05-16 10:49:30,046 controller 57109 application_state.py:202 - Deploy task for app 'MyApp' ran successfully.
+```
+
+Notice that the "Hello from config" message was printed from within the deployment constructor.
```
Comment (angelinalg) [inline doc/source/serve/app-builder-guide.md:74]:
```suggestion
Notice that the "Hello from config" message is printed from within the deployment constructor.
```

File: doc/source/serve/app-builder-guide.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+(typed-app-builders)=
+### Typing arguments with Pydantic
+
+To avoid writing logic to parse and validate the arguments by hand, define a [Pydantic model](https://pydantic-docs.helpmanual.io/usage/models/) as the single input parameter's type to your application builder function (the parameter must be type annotated).
+Arguments are passed the exact same way, but the resulting dictionary is used to construct the Pydantic model using `model.parse_obj(args_dict)`.
```
Comment (angelinalg) [inline doc/source/serve/app-builder-guide.md:80]:
```suggestion
Arguments are passed the same way, but the resulting dictionary is used to construct the Pydantic model using `model.parse_obj(args_dict)`.
```

File: doc/source/serve/app-builder-guide.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+(HTTPProxyActor pid=56828) INFO:     Started server process [56828]
+(ServeReplica:default_HelloWorld pid=56830) Message: Hello from CLI
+2023-05-16 10:47:36,131 SUCC scripts.py:424 -- Deployed Serve app successfully.
+```
+
+## Common patterns
+
+### Multiple parametrized applications using the same builder
+
+Application builders can be used to run multiple applications with the same code but different parameters.
```
Comment (angelinalg) [inline doc/source/serve/app-builder-guide.md:103]:
```suggestion
You can use application builders to run multiple applications with the same code but different parameters.
```

File: doc/source/serve/app-builder-guide.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      model_uri: s3://my_bucket/model_2
+  - name: Model3
+    import_path: my_module:my_model_code
+    args:
+      model_uri: s3://my_bucket/model_3
+```
+
+### Configuring multiple composed deployments
+
+The arguments passed in can be used to configure multiple deployments in a single application.
```
Comment (angelinalg) [inline doc/source/serve/app-builder-guide.md:125]:
```suggestion
You can use the arguments passed in to configure multiple deployments in a single application.
```

**Comment by angelinalg** [comment]
@edoakes, do you want to include this in the side nav? If so, you'll need to add it to the _toc.yml file.

**Review by angelinalg — APPROVED** [review]
LGTM

**Pull request merged**